File: leetcode_java/46. Permutations.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> permute(int[] nums) {
        swap(nums, 0);
        return res;
    }
    private void swap(int[] nums, int j) {
        if (j == nums.length - 1) {
            List<Integer> list = new ArrayList<>();
            for (int num : nums) list.add(num);
            res.add(list);
        }
        for (int i = j; i < nums.length; i++) {
            int tmp = nums[i];
            nums[i] = nums[j]; nums[j] = tmp;
            swap(nums, j + 1);
            nums[j] = nums[i]; nums[i] = tmp;
        }
    }
}

File: leetcode_python/46. Permutations.py
class Solution:
    def permute(self, nums: [int]) -> [[int]]:
        self.res = []
        self.swap(nums, 0)
        return self.res

    def swap(self, nums, j):
        if j == len(nums) - 1: self.res.append(list(nums))
        for i in range(j, len(nums)):
            nums[i], nums[j] = nums[j], nums[i]
            self.swap(nums, j+1)
            nums[i], nums[j] = nums[j], nums[i]
